Implementing QG Leith

edoddridge opened the pull request:
## What changes does this PR introduce?
Implementation of QG Leith subgrid model as described in [Bachman et al. (2017)](https://agupubs.onlinelibrary.wiley.com/doi/abs/10.1002/2016JC012265) and [Pearson et al. (2017)](http://dx.doi.org/10.1016/j.ocemod.2017.05.007).

## Other information:
~~**This branch is not ready to be merged.** It is intended to provide a space for collaboration as we work on the code.~~

The initial code dump has been placed within the `verification` folder as a subtle reminder that we will need a test for this new functionality before it can be merged.

## To do
This list constitutes an initial pass - there is more to do than listed here. Some additional issues are described in 'QG_LEITH_README'
- [x] remove extraneous files
- [x] remove CSV headers
- [x] move subroutines to their correct homes within the source code
- [x] Remove the dynamic computation of the Leith coefficient? Bachman et al. (2017) shows that it isn't particularly useful
- [x] fix the boundary condition issue in `calc_germanoK.F` lines 328-335
- [x] include a verification experiment that tests the new code
- [x] new subroutine `MOM_CALC_QG_STRETCHING` is only called from `mom_fluxform.F` not `mom_vecinv.F`

edoddridge commented:
I've invited @dfer22 and @septicscuzzy to be collaborators on my fork, so that they can push directly to this branch.

edoddridge commented:
These two commits implement a QG version of Leith. They also have the limiter from eqn. (56) of Bachman et al. (2017) that should gracefully transition to 2D Leith as N^2 -> 0.

The code compiles, and turning it on by setting `useLeithQG=.TRUE.` and `viscC2LeithQG = 1.` in `data` for the verification experiment MLAdjust causes testreport to fail - so it is changing something about the solution. The biggest changes are for `u` and `v`, which bodes well, since that's what should change when viscosity changes.

```
default 10  ----T-----  ----S-----  ----U-----  ----V-----  --PTR 01--  --PTR 02--  --PTR 03--  --PTR 04--  --PTR 05--
G D M    c        m  s        m  s        m  s        m  s        m  s        m  s        m  s        m  s        m  s
e p a R  g  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .
n n k u  2  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d
2 d e n  d  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .

Y Y Y Y> 3<16 16 16  8 16 16 16 11  3  3  6  5  4  4  4  4  .  .  .  .  .  .  .  . FAIL  MLAdjust

```

The next step is to provide the option to use the LeithQG coefficient to do some mixing in the GMRedi package.

edoddridge commented:
I ran a couple of toy simulations overnight and the output looks reasonable. The gif below is from a reentrant channel that was initialised with a tube of buoyant fluid at the surface. The instability grows and the viscosity increases as expected. Most of the domain was initially unstratified, so it seems as though this is able to cope with unstratified fluids.

![output](https://user-images.githubusercontent.com/6049169/40247422-26c5d3c0-5a81-11e8-972a-e2312047395c.gif)

The next phase of testing is to put it in some more complex models. I'll try it in a channel with complex forcings, a mixed layer, and sea ice. @septicscuzzy has committed to running one of the simulations from their paper with this code, so that we can compare the outputs.

dfer22 commented:
Ed, can I try it in my channel, the one I used to test the normal 2D Leith? It would tick a lot of the things you mention above.

edoddridge commented:
That would be sensational!

dfer22 commented:
Just to be sure I'm doing the right thing:
I got your code this way:
git clone git@github.com:edoddridge/MITgcm.git
git checkout QG-Leith
and compiled.
Then, in the run, I set:
viscC2LeithQG=1
useLeithQG=1
all other Leith parameters to zero.
My reading of the code is that the divergent term is automatically included in QG Leith.
All good?

edoddridge commented:
Yep, those commands will work and the divergence term is included. `useLeithQG` is a Boolean, and I'm not sure how robust casting fro integers to logicals is across compilers - might be better to use `.TRUE.`?

If you want to use the LeithQG coefficient in GM, you'll need to modify `MOM_COMMON_OPTIONS.h` and define `ALLOW_GM_LEITH_QG`.

edoddridge commented:
I've added two new variants to the `MLAdjust` verification experiment - one uses LeithQG with GMRedi, the other without.

jm-c commented:
I tried testreport with gfortran and -devel and I got a Floating-point exception in mom_visc_qgleith_limiter
in both experiments (AhLeithQG and AhLeithQG_GM).

edoddridge commented:
Thanks @jm-c, I'll look into that.

@septicscuzzy @dfer22 - how did your simulations turn out?

When used with GM, the current implementation doesn't allow for reproducible restarts. Since the viscosity is stored and used in GM on the following timestep, the first step after a restart will use different values. This can be fixed either by storing the viscosity in the pickup files, or calculating it before GM is called. It would be good to keep it out of the pickups, so I'll try calculating earlier in each timestep.

septicscuzzy commented:
Not good, in comparison to my old experiments.  The KE spectra have ski
jumps.

However, I ran with useFullLeith = True on these, while it was False on the
old simulations.  I'll try switching to false and see if that changes
anything.


On Mon, Jun 4, 2018 at 8:47 AM, Ed Doddridge <notifications@github.com>
wrote:

> Thanks @jm-c <https://github.com/jm-c>, I'll look into that.
>
> @septicscuzzy <https://github.com/septicscuzzy> @dfer22
> <https://github.com/dfer22> - how did your simulations turn out?
>
> When used with GM, the current implementation doesn't allow for
> reproducible restarts. Since the viscosity is stored and used in GM on the
> following timestep, the first step after a restart will use different
> values. This can be fixed either by storing the viscosity in the pickup
> files, or calculating it before GM is called. It would be good to keep it
> out of the pickups, so I'll try calculating earlier in each timestep.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MITgcm/MITgcm/pull/96#issuecomment-394381145>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ATrzIjWJAfPTQ_9PSv0paaT4hiNJEhJ9ks5t5UhlgaJpZM4Txn35>
> .
>

edoddridge commented:
Ah, that's not ideal. That'd be great, thanks.

septicscuzzy commented:
Well, if there was ever any question about whether UseFullLeith matters,
here is the answer:

I think these are close enough to the spectra from my original paper to be
satisfactory, given the numerous small changes that have been made (i.e.
limiters, fixing vorticity on the boundary, etc.).  I would definitely add
a warning to *not* use Full Leith with QG Leith, though.

On Mon, Jun 4, 2018 at 9:35 AM, Ed Doddridge <notifications@github.com>
wrote:

> Ah, that's not ideal. That'd be great, thanks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MITgcm/MITgcm/pull/96#issuecomment-394399152>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ATrzIlFgG4VY8cXN6vQAaavIwElKn26Yks5t5VPEgaJpZM4Txn35>
> .
>

edoddridge commented:
I haven't seen them - did you mean to attach them to your comments?

septicscuzzy commented:
I was referring to the spectra in the attachment on my June 5 email.  You
asked for "nice versions", but what does that mean?

On Thu, Jun 7, 2018 at 8:12 AM, Ed Doddridge <notifications@github.com>
wrote:

> I haven't seen them - did you mean to attach them to your comments?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MITgcm/MITgcm/pull/96#issuecomment-395434755>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ATrzIqNM2YL8iLeST4W8utojAjfSBs0yks5t6TS0gaJpZM4Txn35>
> .
>

edoddridge commented:
I'm pretty sure that attachments are ignored by GitHub when commenting via email - certainly nothing showed up here on the thread.

edoddridge commented:
Apart from updating `tag-index` I think this is done. I'll leave that until it's been reviewed to avoid potential conflicts.

- the floating point exception that @jm-c pointed out before has been fixed: `testreport` passes with the `-devel` option
- the new functionality has been documented, including the runtime and compile time options, as well as a recommended value for the non-dimensional viscosity parameter
- both locations in which the QGLeith viscosity is calculated (`pkg/gmredi` and `pkg/mom_common`) can output the array through `pkg/diagnostics` (and the values are extremely similar)
- the GM functionality is now controlled by both compile time and runtime parameters, so it can be switched on and off at will

jm-c commented:
Good news: I checked that the MLAdjust examples pass test 2+2 ; also checked that results 
  don't change when I #define DYNAMICS_GUGV_EXCH_CHECK in dynamics.F, tested on CS-grid 
  with dry corners.
Regarding code changes:
1) currently, serious problems with Adjoint (I only tried TAF, none of verif exp can
  even compile); We need a specific CPP option to protect this code (and 2 new calls 
  in mom_fluxform.F and mom_vecinv.F) from being always seen by TAF. 
2) option ALLOW_GM_LEITH_QG  is set in 2 places and it's confusing. Since it's only 
  used in pkg/gmredi it should stay there and should be removed from MOM_COMMON_OPTIONS.h
3) what happens if I try: viscC2LeithQG=1 and useLeithQG=F ? 
  seems that results are different from full QG-Leith (=1,=T) but also different
  from no QG-Leith at all (=0,=F).
  Seems that in some part of the code it tests for viscC2LeithQG > 0 and in other places 
  it tests for useLeithQG, but would be better to to clarify this.
4) few "#include "MOM_VISC.h" have been added in pkg/gmredi but I am not convinced
  that all are necessary. Also the 3 added *.F files could have shorter name, currently
  mom_visc_qgleith_calc_stretching.F is the longest of any pkg, and all in lower-case
  (except suffix .F) is preferable.

On Thu, Jul 12, 2018 at 02:16:09PM -0700, Ed Doddridge wrote:
> Apart from updating `tag-index` I think this is done. I'll leave that until it's been reviewed to avoid potential conflicts.
> 
> - the floating point exception that @jm-c pointed out before has been fixed: `testreport` passes with the `-devel` option
> - the new functionality has been documented, including the runtime and compile time options, as well as a recommended value for the non-dimensional viscosity parameter
> - both locations in which the QGLeith viscosity is calculated (`pkg/gmredi` and `pkg/mom_common`) can output the array through `pkg/diagnostics` (and the values are extremely similar)
> - the GM functionality is now controlled by both compile time and runtime parameters, so it can be switched on and off at will
> 
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/MITgcm/MITgcm/pull/96#issuecomment-404653335

edoddridge commented:
Thanks @jm-c.

I've implemented all of your suggestions. 

1) I haven't tested the adjoint, but I think I've managed to protect the new code with the `ALLOW_LEITH_QG` and `ALLOW_GM_LEITH_QG` CPP flags.

2) `ALLOW_GM_LEITH_QG` was only meant to be in `GMREDI_OPTIONS.h`, the copy in `MOM_OPTIONS.h` was an unintentional remnant. Thanks for catching it.

3) I removed the runtime logical parameter `useLeithQG` - there's now an internal equivalent that is set depending on the value of `viscC2LeithQG `. 

4) Superfluous includes have been flushed.
4.1) File and subroutine names have been shortened.

@jrscott - I've documented these parameters in chapter 2, but they should also be added to the full list of parameters and cpp flags in the future.

jrscott commented:
yes, suggest we check it PR#123 (soon!) and then add the additional changes to the doc.

Also, can we make this more specific:
LeithQG viscosity is designed to work best in simulations that resolve some mesoscale features. In simulations that are too coarse to permit eddies or fine enough to resolve submesoscale features, it should fail gracefully. 

sounds like this might be an option for eddy-permitting res runs that might still also use GM?  (i.e. mesoscale, "fail gracefully" are too vague). Be more specific in recommendation on when this might be a good option.

jm-c commented:
I run testreport but tutorial_cfc_offline failed to compile (gmredi_calc_qgleith.F). I also have one
adjoint experiment that fail to compile (tutorial_dic_adjoffline) but this could be for the same reason.

jm-c commented:
It's getting better, and the AD experiments are now running (except 
tutorial_dic_adjoffline, cf previous comment) but some AD results have 
changed a bit: 
< Y Y Y Y 16>12<16 pass  1D_ocean_ice_column  (e=0, w=3)
< Y Y Y Y 16>12<16 pass  bottom_ctrl_5x5  (e=0, w=7)
< Y Y Y Y 16>13<16 pass  global_ocean.90x40x15  (e=0, w=15)
---
< Y Y Y Y 16>13<16 pass  halfpipe_streamice  (e=0, w=10)
---
< Y Y Y Y 16>13<16 pass  lab_sea.noseaicedyn
---
< Y Y N N .. .. .. N/O   tutorial_dic_adjoffline  (e=0, w=1)
< Y Y Y Y 14>13< 5 pass  tutorial_global_oce_biogeo  (e=0, w=6)
compared to previously:
> Y Y Y Y 16>16<16 pass  1D_ocean_ice_column  (e=0, w=3)
> Y Y Y Y 16>16<16 pass  bottom_ctrl_5x5  (e=0, w=7)
> Y Y Y Y 16>16<16 pass  global_ocean.90x40x15  (e=0, w=15)
---
> Y Y Y Y 16>16<16 pass  halfpipe_streamice  (e=0, w=10)
---
> Y Y Y Y 16>16<16 pass  lab_sea.noseaicedyn
---
> Y Y Y Y 16>16< 8 pass  tutorial_dic_adjoffline  (e=0, w=1)
> Y Y Y Y 16>16<15 pass  tutorial_global_oce_biogeo  (e=0, w=6)
will need to investigate why it's like this.

Otherwise, I took a closer look to the code and have few comments:
Regarding coding:
1) the new subroutines: all the arguments should be documented inside,
  and it does not hurt to add the usual CBOP / CEOP + standard format comments
2) it's really useful for debugging to have myTime and especially myIter 
  available withing the S/R. So i would suggest to add them as argument.
3) probably better to use 3. _d 0 and 4. _d 0 instead of just 3 and 4
  in mom_visc_qgl_stretch.F (the only exception would be the power)
  Note that these simple constant: zeroRL, oneRL, twoRL, halfRL
  are defined in EEPARAMS.h and therefore available everywhere.
4) the "1. _d -9" constant is used in several places, so it might be better
  to use a named constant for it (like QGL_epsil) in case one want to 
  change it (or test if it does make a difference).
 Also, in mom_visc_qgl_limit.F, line 76, the "-9" is beyond 72 column.
 Luckily, after CPP processing it got moved inside the 72.c so it's OK
 but little bit risky.
5) in gmredi_calc_tensor.F, line 295, this will reset array GM_LeithQG_K
 to zero nSx*nSy times (since it's within bi,bj loops). Not great and 
 could be confusing when running multi-threaded.

Algorithm:
1) main issue: the special code for k=1 and k=Nr is a problem, because
  the bottom is not always at k=1 (in p-coords) or at k=Nr (in z-coords).
  In addition, below an ice-shelf, the uppermost grid cell is not at k=1 in 
  z-coords (p-coords is not an issue since it's not working with pkg/shelfice).
 I would argue that:
  a) either this special treatment is really necessary, and in this case
   it needs to be implemented for non-uniform kLowC (=Nr) and non-uniform
   kSurfC (=1) 
  b) or this special treatment is not essential and current code give
   good results even above kLowC < Nr, and in this case we should remove
   the piece of code dealing with k=1 and k=Nr.
 The idea is that the results are not expected to change if I increase
  Nr by 1 and fill the bottom level with land.
2) In many places, computation is not right because it ignores the precise
   location of variables on C-grid ; but since the exact location is not 
   clearly documented, I might have missed some.
  a) computation of N2_2 (line 400 and above), should be recip_drC(k+1) ;
    and not convinced that the right recip_dr is used in computation of N2_3 neither.
  b) horizontal location of "stretching" is at grid-cell center ; this is
   not consistent with lines 87-88 of mom_visc_qgl_limit.F since vort3 is
   at grid-cell corner. Not consistent neither with computation of vrtDx & vrtDy
   in mom_calc_visc.F (assumed to be at same location as vort3).
  c) GM_LeithQG_K seems to be at grid-cell center (where Theta & Salt are) 
    but vertical averaging is missing when computing Kwx, Kwy and Kwz (in
    gmredi_calc_tensor.F). 
3) GM_useLeithQG is not implemented in gmredi_calc_psi_b.F ; however, it contributes
  to the GM part when using the skew flux. This is a little bit strange.
  Also, would it be difficult to implement a run-time switch to get QG-Leith only 
  for Redi or only for GM ?
  And an open question: is there any need to implement GM_LeithQG_K in 
  gmredi_calc_psi_bvp.F ?
4) the code is not always safe regarding the risk of division by zero:
  a) in mom_visc_qgl_stretch.F, could make sure that N2 is not negative,
    using: MIN( zeroRL, sigmaR ) instead of just sigmaR
  b) in mom_visc_qgl_limit.F, will not work when a grid-cell is centered on 
   the equator, because of division by fCori(i,j,bi,bj)**2 ; since the sign 
   does not matter, should be easy to fix using a max.
5) last thing: will need to add some check and stop but can consider this
  later on.

edoddridge commented:
Thanks @jm-c. I'm going to be working on this again in the next couple of weeks.

@septicscuzzy - do you have any comments on these?

septicscuzzy commented:
Using limiters on N2 and fCori to prevent division by zero is entirely
sensible -- the QG theory breaks down at the equator and when the
stratification is really weak anyways, so we just need to handle these
situations in a minimally intrusive way.

On Mon, Sep 10, 2018 at 9:31 AM, Ed Doddridge <notifications@github.com>
wrote:

> Thanks @jm-c <https://github.com/jm-c>. I'm going to be working on this
> again in the next couple of weeks.
>
> @septicscuzzy <https://github.com/septicscuzzy> - do you have any
> comments on these?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MITgcm/MITgcm/pull/96#issuecomment-419956021>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ATrzIujUp61aBXgo_trpo8b4uqUQ8tzmks5uZoXhgaJpZM4Txn35>
> .
>

edoddridge commented:
@jm-c - I've been implementing your suggestions for tidying the code up. Mostly done, apart from the grid locations, which will take a bit longer.

> the "1. _d -9" constant is used in several places, so it might be better
to use a named constant for it (like QGL_epsil) in case one want to
change it (or test if it does make a difference).

Were you imagining that this would be a proper runtime parameter that comes through the input files and has a default, or just a local variable that could be changed prior to compiling. I'm happy to implement it with all the bells and whistles, I'm just not sure how much use it would get.

@septicscuzzy - Cheers. I agree. 

Do you have strong feelings or insights on @jm-c's more substantial comments?

Removing the special code for `k=1` or `k=Nr`
- Thoughts?
- At the moment the main section in `mom_visc_qgl_stretch.F` calculates the vertical derivative as centred differences. With some clever masking we could use the exact same code to calculate a one sided difference at the top and bottom. 

Having this available for only Redi or only GM?
- My understanding is that adding QGLeith to GM is just a nice way of smoothing the fields, so it should work just as well for either Redi or GM. Am I missing something fundamental?

septicscuzzy commented:
I'm glad jm-c looked through it in detail. I'm not surprised at most of his
points, as I was expecting some degree of slop in my original coding that
would need to be sorted out before the module became canon.  The details
regarding the "k=1, k=Nr" problem were things I was not aware of.  I think
doing one-sided differences is a good solution.  Regarding Redi/GM, the
theory is only validated for GM and we never explored what would happen if
you made it available to Redi too.  I don't think it would be a bad idea to
allow Redi = QGL too if that is what the user wants, with the caveat that
any effects from such a choice are unknown and undocumented.

On Mon, Sep 24, 2018 at 1:21 PM Ed Doddridge <notifications@github.com>
wrote:

> @jm-c <https://github.com/jm-c> - I've been implementing your suggestions
> for tidying the code up. Mostly done, apart from the grid locations, which
> will take a bit longer.
>
> the "1. _d -9" constant is used in several places, so it might be better
> to use a named constant for it (like QGL_epsil) in case one want to
> change it (or test if it does make a difference).
>
> Were you imagining that this would be a proper runtime parameter that
> comes through the input files and has a default, or just a local variable
> that could be changed prior to compiling. I'm happy to implement it with
> all the bells and whistles, I'm just not sure how much use it would get.
>
> @septicscuzzy <https://github.com/septicscuzzy> - Cheers. I agree.
>
> Do you have strong feelings or insights on @jm-c <https://github.com/jm-c>'s
> more substantial comments?
>
> Removing the special code for k=1 or k=Nr
>
>    - Thoughts?
>    - At the moment the main section in mom_visc_qgl_stretch.F calculates
>    the vertical derivative as centred differences. With some clever masking we
>    could use the exact same code to calculate a one sided difference at the
>    top and bottom.
>
> Having this available for only Redi or only GM?
>
>    - My understanding is that adding QGLeith to GM is just a nice way of
>    smoothing the fields, so it should work just as well for either Redi or GM.
>    Am I missing something fundamental?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MITgcm/MITgcm/pull/96#issuecomment-424093882>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ATrzIkRbSgIkmV1ThYPX5rWKDqv9-x5jks5ueTDVgaJpZM4Txn35>
> .
>

edoddridge commented:
Next iteration:

> main issue: the special code for k=1 and k=Nr is a problem, because
> the bottom is not always at k=1 (in p-coords) or at k=Nr (in z-coords).
> In addition, below an ice-shelf, the uppermost grid cell is not at k=1 in
> z-coords (p-coords is not an issue since it's not working with pkg/shelfice).
> I would argue that:
> a) either this special treatment is really necessary, and in this case
> it needs to be implemented for non-uniform kLowC (=Nr) and non-uniform
> kSurfC (=1)
> b) or this special treatment is not essential and current code give
> good results even above kLowC < Nr, and in this case we should remove
> the piece of code dealing with k=1 and k=Nr.
> The idea is that the results are not expected to change if I increase
> Nr by 1 and fill the bottom level with land.

Agreed. I've removed the special computations for `k=1` and `k=Nr`. There is instead a series of `IF` statements that determines if we're at the top, in the middle, or at the bottom and sets some offsets to appropriate values. These offsets are then used within the main code so that it can handle each situation. I didn't end up using one sided differencing, because I needed to difference the slopes. In this implementation the top level gets the same viscosity as the one below it, and the bottom level the same as the one above it. This is a small change from @septicscuzzy's implementation where they got the same value as the one two below or above - is there a theoretical argument I'm missing that makes that a better choice?

One quirk is that the algorithm requires density above and below the level being evaluated in order to evaluate vertical derivatives of stratification. Obviously, this isn't possible at the top or the bottom. Rather than always calculate density at five separate levels (the level of the `k` loop, plus the ones either side for normal situations, and the ones immediately above and below those for the top and bottom cases) I've included calls to `FIND_RHO_2D` that return just the density in the required grid cell when the `IF` tree shows that we need density at an additional level. 


> In many places, computation is not right because it ignores the precise
> location of variables on C-grid ; but since the exact location is not
> clearly documented, I might have missed some.
> a) computation of N2_2 (line 400 and above), should be recip_drC(k+1) ;

Yep. Fixed.

> and not convinced that the right recip_dr is used in computation of N2_3 neither.

Yep. This one needs the sum of `drC(k)` and `drC(k+1)`.

> b) horizontal location of "stretching" is at grid-cell center ; this is
> not consistent with lines 87-88 of mom_visc_qgl_limit.F since vort3 is
> at grid-cell corner. 

I've averaged `vort3` to the cell centres so that it's at the same place as the other terms used to limit `stretching`.

> Not consistent neither with computation of vrtDx & vrtDy
> in mom_calc_visc.F (assumed to be at same location as vort3).

I think `vrtDx` is at the v-points and `vrtDy` is at the u-points. I've altered the QGLeith code so that the gradients of `stretching are averaged onto these same points.

> c) GM_LeithQG_K seems to be at grid-cell center (where Theta & Salt are)
> but vertical averaging is missing when computing Kwx, Kwy and Kwz (in
> gmredi_calc_tensor.F).

Indeed. `GM_LeithQG_K` is now vertically averaged for `Kwx`, `Kwy`, and `Kwz`.

> GM_useLeithQG is not implemented in gmredi_calc_psi_b.F ; however, it contributes
> to the GM part when using the skew flux. This is a little bit strange.

Good point. I've included it in `gmredi_calc_psi_b.F`. I had to also pass `myTime` and `myIter` to `GMREDI_CALC_PSI_B` since `GMREDI_CALC_QGLEITH` requires them as arguments.

> Also, would it be difficult to implement a run-time switch to get QG-Leith only
> for Redi or only for GM ?

Given @septicscuzzy's comment:
> > Regarding Redi/GM, the theory is only validated for GM and we never explored what would happen if you made it available to Redi too.  I don't think it would be a bad idea to allow Redi = QGL too if that is what the user wants, with the caveat that any effects from such a choice are unknown and undocumented.

It doesn't seem like a great idea to just code it up for the sake of coding it - if we have no idea what it will do, or whether it's a good idea to do it, why put it in?

> And an open question: is there any need to implement GM_LeithQG_K in
> gmredi_calc_psi_bvp.F ?

I've never used that routine, but from memory, it doesn't allow for variable kappa_GM schemes like Visbeck, does it? So, probably no need to implement this?

> the code is not always safe regarding the risk of division by zero:
> a) in mom_visc_qgl_stretch.F, could make sure that N2 is not negative,
> using: MIN( zeroRL, sigmaR ) instead of just sigmaR
> b) in mom_visc_qgl_limit.F, will not work when a grid-cell is centered on
> the equator, because of division by fCori(i,j,bi,bj)**2 ; since the sign
> does not matter, should be easy to fix using a max.

I've implemented both of those. 

> last thing: will need to add some check and stop but can consider this
> later on.

OK, let's chat.

Still to do:
- [ ] There's a floating point exception (IEEE_INVALID_FLAG) that I haven't tracked down yet.
- [ ] These changes alter the results of the test simulations. Given that I've shifted variables on the grid and altered the algorithm, that's expected, but requires validation. We need to run some bigger simulations and analyse the output.

jm-c commented:
@edoddridge going to talk to you about way to skip recomputing density (call to EOS).

also noticed:
a) myTime is declared INTEGER but  should be _RL
b) in mom_calc_visc.F, line 437, leithQG2fac should be inside SQRT function.

In general, thecurrent  implementation of QG-Leith does not zero out gradient between land point
and ocean point. 
1) When computing stretching term,  if should be possible to make sure the stretching is right
 near bottom & surface for the case of uniform stratification.
2) regarding horizontal gradient of stretching, since stencil is much wider that original one (with only
grad of vorticity involved), might be better to zero out (multiply by maskW & maskS) the gradient
that involve land point.

edoddridge commented:
@jm-c - these latest commits avoid recomputing density by using `sigmaRfield` from `DYNVARS.h`. The storage is protected by cpp options.

When dealing with vertical locations, the edge cases are dealt with more elegantly now:
- one level of ocean: `stretching` is set to zero
- two levels of ocean, in the upper level: use stratification from lower face of tracer cell, assume it is constant in the vertical
- two levels of ocean, in the lower level: use stratification from the upper face of tracer cell, assume it is constant in the vertical


Regarding masking the horizontal gradients, the gradient of `stretching` is set to zero based on `maskS`:

https://github.com/MITgcm/MITgcm/blob/ab507001dda4f4ddb7a52a2ff6598c5853717efc/pkg/mom_common/mom_calc_visc.F#L333-L343

and `maskW`:

https://github.com/MITgcm/MITgcm/blob/ab507001dda4f4ddb7a52a2ff6598c5853717efc/pkg/mom_common/mom_calc_visc.F#L353-L365

(sorry for all the noise on this branch)

jm-c commented:
@edoddridge I started to check the changes/addition to pkg/gmredi ; I saw some problems so I made
few adjustments there. I might do other minor update to reduce number of difference with original code
(since it helps to find problems). I have few suggestions:
1) we don't need to support QG-Leith in GM for the case where "GM_NON_UNITY_DIAGONAL" is undef;
 Just adding a "stop" in gmredi_check.F for this case should do it; with this restriction, we don't need to change gmredi_xtransport.F &gmredi_ytransport.F  (and BTW, we need also a stop when GM_useLeithQG=T but ALLOW_GM_LEITH_QG is undef).
2) shorter experiments name ? I would drop "Ah" since there is no bi-harmonic version;
 maybe: MLAdjust.QGLeith and MLAdjust.GQLthGM ?
3) maybe there is a problem in: MLAdjust/input.AhLeithQG_GM/data.gmredi
 with GM_useBVP=T since I don't see QG-Leith being implemented in BVP routine.

jm-c commented:
@edoddridge I think there is still one remaining problem: Nsquare from output of S/R mom_visc_qgl_stretch.F is at vertical interface k, between level k-1 and level k. But inside S/R mom_visc_qgl_limit.F, it is assumed to be at level center k. May be simplest would be to vertically average Nsquare inside mom_visc_qgl_stretch.F (with all 3 cases) ?
And 2 comments on the side: 
a) it might be possible to simplify a bit mom_visc_qgl_limit.F, skipping SQRT function and removing
few **2 or **4 calculation.
b) related to above: might be good to explicitly add description of the grid location of new variables
GM_LeithQG_K, Nsquare, streching (+ may be few others that I might missed).
Currently re-running full set of ADM & FWD experiments.

jm-c commented:
@edoddridge there are 2 experiments (1 fwd, 1 adm) that do not compile, likely because of my changes. Going to check this.

jm-c commented:
Should be better now: both experiments tutorial_cfc_offline and tutorial_dic_adjoffline are passing. Have not 
checked  the OpenAD testreport but will wait until final update are done.

edoddridge commented:
Thanks @jm-c.

I rolled back the changes in `gmredi_xtransport.F` and `gmredi_ytransport.F` and put in a stop for when `GM_NON_UNITY_DIAGONAL` is undef.

I also put in a `STOP` for the case in which `ALLOW_GM_LEITH_QG` is undef and `GM_useLeithQG=T`, but the model doesn't get that far. It hits a `STOP` in pkg/diagnostics because `GM_LTHQG` isn't defined:

```
runmodel in MLAdjust/tr_run.QGLthGM ... failed (run: 0  end: 0 )
=> output from running in MLAdjust/tr_run.QGLthGM :
>  link mitgcmuv from dir ../build
> STOP ABNORMAL END: S/R DIAGNOSTICS_SET_POINTERS
> DIAGNOSTICS_SET_POINTERS: GM_LTHQG is not a Diagnostic
```

I dropped the "Ah" from the verification experiments and shortened the names. Good suggestion.

You're right - I shouldn't have used the BVP formulation of GM. QGLeith doesn't work with that form. It's fixed now.

I averaged `Nsquare` in the first case, but didn't for either of the top or bottom scenarios. At the top and the bottom, I made the algorithm assume the stratification is constant, so I just made sure that the array `Nsquare` gets given the right value.

Good point about `mom_visc_qgl_limit.F`. I removed the `SQRT` and carried around the squared value, which removes some computation.

I added some comments about grid location of the new variables. 

I realised that I hadn't averaged the stretching gradient correctly in `gmredi_calc_qg_leith.F`. That's been fixed.

I'm currently running testreport (forward experiments only) on my laptop.

edoddridge commented:
All the forward tests compile and run. There are a few where the output differs enough from the stored outputs that I get  a "FAIL", but those simulations also fail when I run testreport on `master`, so it's not related to these changes.

I'm going to merge `master` into this branch so I can add the new parameters to the [table in chapter 3](https://mitgcm.readthedocs.io/en/latest/getting_started/getting_started.html#general-dissipation-scheme).

jm-c commented:
@edoddridge I checked that all OpenAD experiments are unaffected (after fixing the little syntax error).
Regarding the 2 new experiment output, there is one annoying thing is that on our reference platform
(baudelaire) results are little bit different (11 digits are similar) ; so it means we will need to get a new version from baudelaire itself. 
And also one comment regarding: 7b627fde0ee0da4e630adeb31c3b3b1aa2a4d65e 
the magnitude of QGL_lim_epsil has not been changed (now used with squared of N2*dz ).  It might be better to put a smaller value (square of current one ?) since typical value of N2 or f are not that large.

edoddridge commented:
That's annoying - I guess you'll need to commit the output from baudelaire.

Two minor fixes:
- reduce `QGL_epsil` to the square of the previous value
- I implemented the same fix as e3b891cdca00e73ef51f326fe648e7894748f05d in the GM subroutine.

jm-c commented:
Forgot to comment after updating output: Everything looks good.
LGTM +1

edoddridge commented:
Thanks @septicscuzzy 

@jrscott - does this mean that the docs have your blessing?

jrscott commented:
thx @septicscuzzy, that makes more sense, fixed.
